Firewall config (#1016)
stackhpc/stackhpc-kayobe-config: doc/source/configuration/firewall.rst

.. _firewall:

========
Firewall
========

StackHPC Kayobe configuration provides a standardised firewalld configuration.
The configuration uses the :kayobe-doc:`firewall
<configuration/reference/hosts.html#firewalld>` host configuration
functionality of Kayobe.

The firewall configuration is provided in
``etc/kayobe/inventory/group_vars/all/firewall``.

Enabling StackHPC firewalld rules
=================================

> The standardised firewalld configuration is not used by default and must be
> actively opted into. This can be done as follows:

Reviewer: ```suggestion
The standardised firewalld configuration is not enabled by default and must be
actively opted into. This can be done by altering or copying ``etc/kayobe/<group>.yml`` to the chosen ``etc/kayobe/environment/<environment>/``, and making the appropriate following changes:
```

Author: Some deployments don't use kayobe environments at all. I think I'll just remove the path from the file caption, since it doesn't matter whether it's in an env or not

Reviewer: ```suggestion
The standardised firewalld configuration is not enabled by default and must be
actively opted into. This can be done by altering or, in the case of using a ``kayobe`` environment, copying ``etc/kayobe/<group>.yml`` to the chosen ``etc/kayobe/environment/<environment>/``, and making the appropriate following changes:
```

Seems odd that we reference paths and files in terms of kayobe environments later on, but not for this section, when it seems equally valid to do so.

Author: I've updated the line to mention both environments and the base config

> Controller firewalld Configuration

Reviewer: ```suggestion
Controller firewalld Configuration - ``etc/kayobe/controllers.yml``
```

Author: This is the title of the subsection. I don't think it needs the path in it since there's already a caption on the code block

----------------------------------

.. code-block:: yaml
   :caption: ``etc/kayobe/controllers.yml``

   ###############################################################################
   # Controller node firewalld configuration.

   # Whether to install and enable firewalld.
   controller_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   controller_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   controller_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   controller_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

> Compute firewalld Configuration

Reviewer: ```suggestion
Compute firewalld Configuration - ``etc/kayobe/compute.yml``
```

-------------------------------

.. code-block:: yaml
   :caption: ``etc/kayobe/compute.yml``

   ###############################################################################
   # Compute node firewalld configuration.

   # Whether to install and enable firewalld.
   compute_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   compute_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   compute_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   compute_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

> Storage firewalld Configuration

Reviewer: ```suggestion
Storage firewalld Configuration - ``etc/kayobe/storage.yml``
```

-------------------------------

.. code-block:: yaml
   :caption: ``etc/kayobe/storage.yml``

   ###############################################################################
   # storage node firewalld configuration.

   # Whether to install and enable firewalld.
   storage_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   storage_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   storage_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   storage_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

> Monitoring firewalld Configuration

Reviewer: ```suggestion
Monitoring firewalld Configuration - ``etc/kayobe/monitoring.yml``
```

----------------------------------

.. code-block:: yaml
   :caption: ``etc/kayobe/monitoring.yml``

   ###############################################################################
   # monitoring node firewalld configuration.

   # Whether to install and enable firewalld.
   monitoring_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   monitoring_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   monitoring_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   monitoring_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

> Infrastructure VM firewalld Configuration

Reviewer: ```suggestion
Infrastructure VM firewalld Configuration - ``etc/kayobe/infra_vms.yml``
```

-----------------------------------------

The standard firewalld configuration has rules for wazuh-manager and Ansible
control host Infrastructure VMs.

.. code-block:: yaml
   :caption: ``etc/kayobe/infra-vms.yml``

   ###############################################################################
   # Infrastructure VM node firewalld configuration

   # Whether to install and enable firewalld.
   infra_vm_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   infra_vm_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   infra_vm_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   infra_vm_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

> Seed firewalld Configuration

Reviewer: ```suggestion
Seed firewalld Configuration - ``etc/kayobe/seed.yml``
```

----------------------------

.. code-block:: yaml
   :caption: ``etc/kayobe/seed.yml``

   ###############################################################################
   # seed node firewalld configuration.

   # Whether to install and enable firewalld.
   seed_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   seed_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   seed_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   seed_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

> Seed Hypervisor firewalld Configuration

Reviewer: ```suggestion
Seed Hypervisor firewalld Configuration - ``etc/kayobe/seed-hypervisor.yml``
```

---------------------------------------

.. code-block:: yaml
   :caption: ``etc/kayobe/seed_hypervisor.yml``

   ###############################################################################
   # seed_hypervisor node firewalld configuration.

   # Whether to install and enable firewalld.
   seed_hypervisor_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   seed_hypervisor_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   seed_hypervisor_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   seed_hypervisor_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

Custom rules
------------

> Custom firewalld rules can be added for any of the following groups using their

Reviewer: ```suggestion
Custom firewalld rules can be added for any of the following groups within an environment's ``group_vars`` inventory, ``etc/kayobe/environments/<environment>/inventory/group_vars/<group>/firewall`` and using their
```

Author: They don't actually have to be in group vars. Host vars, group vars, and extra vars are all valid. I'll update the doc to make that clearer

Author: Done

* Seed Hypervisor - ``stackhpc_seed_hypervisor_firewalld_rules_extra``

Each variable is a list of firewall rules to apply. Each item is a dict
containing arguments to pass to the firewalld module.

The example below would enable SSH on the ``provision_oc`` network, and disable
UDP port 1000 on the ``admin_oc`` network for the Wazuh manager Infrastructure
VM:

.. code-block:: yaml
   :caption: ``etc/kayobe/inventory/group_vars/wazuh_manager/firewall``

   stackhpc_wazuh_manager_infra_vm_firewalld_rules_extra:
     -  service: ssh
        network: "{{ provision_oc_net_name }}"
        zone: "{{ provision_oc_net_name | net_zone }}"
        state: enabled
     -  port: 1000/udp
        network: "{{ admin_oc_net_name }}"
        zone: "{{ admin_oc_net_name | net_zone }}"
        state: disabled

Beware that if any rules are found that directly conflict (a service or port is
both enabled and disabled) the configuration will fail. There is currently no
way to override rules in the standard configuration, other than to find the
rule and delete it manually. If you find a standard rule that does not work for
your deployment, please consider merging your changes back in to upstream SKC.

> Applying changes
> ----------------
> 
> The ``kayobe configuration dump`` command can be used to view all the rules
> that will be applied to a host.
> 
> .. code-block:: bash
> 
>    kayobe configuration dump --var-name stackhpc_firewalld_rules --limit <host>
> 
> If the command above prints a template, rather than a clean list of rules, the
> configuration is invalid. The kayobe configuration dump command can be used on
> other variables such as ``stackhpc_firewalld_rules_unverified`` or
> ``stackhpc_*_firewalld_rules`` to debug the configuration. See the `How it
> works`_ section for more details.
> 
> Ensure Kolla Ansible opens up ports in firewalld for services on the public
> API network:
> 
> .. code-block:: yaml
>    :caption: ``etc/kayobe/kolla/globals.yml``
> 
>    enable_external_api_firewalld: true
>    external_api_firewalld_zone: "{{ public_net_name | net_zone }}"
> 
> Ensure every network in ``networks.yml`` has a zone defined. The standard
> configuration is to set the internal network zone to ``trusted`` and every
> other zone to the name of the network. See
> ``etc/kayobe/environments/ci-multinode/networks.yml`` for a practical example.

Reviewer: ```suggestion
Prerequisites
----------------

Ensure Kolla Ansible opens up ports in firewalld for services on the public
API network:

.. code-block:: yaml
   :caption: ``etc/kayobe/kolla/globals.yml``

   enable_external_api_firewalld: true
   external_api_firewalld_zone: "{{ public_net_name | net_zone }}"

Ensure every network in ``networks.yml`` has a zone defined. The standard
configuration is to set the internal network zone to ``trusted`` and every
other zone to the name of the network. See
``etc/kayobe/environments/ci-multinode/networks.yml`` for a practical example.

Applying changes
----------------

The ``kayobe configuration dump`` command can be used to view all the rules
that will be applied to a host.

.. code-block:: bash

   kayobe configuration dump --var-name stackhpc_firewalld_rules --limit <host>

If the command above prints a template, rather than a clean list of rules, the
configuration is invalid. The kayobe configuration dump command can be used on
other variables such as ``stackhpc_firewalld_rules_unverified`` or
``stackhpc_*_firewalld_rules`` to debug the configuration. See the `How it
works`_ section for more details.
```

Author: Good point. I've gone one step further and split out the KA config and config validation

.. _firewall:

========
Firewall
========

StackHPC Kayobe configuration provides a standardised firewalld configuration.
The configuration uses the :kayobe-doc:`firewall
<configuration/reference/hosts.html#firewalld>` host configuration
functionality of Kayobe.

The firewall configuration is provided in
``etc/kayobe/inventory/group_vars/all/firewall``. This allows configuration
variables to be overridden on a per-group or per-host basis (which would not be
possible for an "extra variable" in ``etc/kayobe/*.yml``). This configuration
is not used by default, and must be actively opted into. This can be done as
follows:

.. code-block:: yaml
   :caption: ``etc/kayobe/controllers.yml``

>    controller_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

Reviewer 2: Better add the enable flag and default zone here too

Author: Done

   controller_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

This will configure the standard set of firewalld rules on controller hosts.
Rule definitions are automatically added according to group membership.Rule
sets exist for the following groups:

* Controllers - ``stackhpc_controller_firewalld_rules``
* Compute - ``stackhpc_compute_firewalld_rules``
* Storage - ``stackhpc_storage_firewalld_rules``
* Monitoring - ``stackhpc_monitoring_firewalld_rules``
* Wazuh Manager Infrastructure VM - ``stackhpc_infra_vm_firewalld_rules``

> * Ansible Control host Infrastructure VM - ``stackhpc_infra_vm_firewalld_rules``

Reviewer 2: Wrong variable names

Author: Done

is not used by default, and must be actively opted into. This can be done as
follows:

.. code-block:: yaml
   :caption: ``etc/kayobe/controllers.yml``

   controller_firewalld_enabled: true
   controller_firewalld_rules: "{{ stackhpc_firewalld_rules }}"
   controller_firewalld_zones: "{{ stackhpc_firewalld_zones }}"
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   # Unset to leave the default zone unchanged
   controller_firewalld_default_zone: drop

This will configure the standard set of firewalld rules on controller hosts.
Rule definitions are automatically added according to group membership. Rule
sets exist for the following groups:

* Controllers - ``stackhpc_controller_firewalld_rules``
* Compute - ``stackhpc_compute_firewalld_rules``
* Storage - ``stackhpc_storage_firewalld_rules``
* Monitoring - ``stackhpc_monitoring_firewalld_rules``
* Wazuh Manager Infrastructure VM - ``stackhpc_wazuh_manager_infra_vm_firewalld_rules``
* Ansible Control host Infrastructure VM - ``stackhpc_ansible_control_infra_vm_firewalld_rules``
* Seed - ``stackhpc_seed_firewalld_rules``

> * Seed Hypervisor - ``stackhpc_seed_hypervisor_firewalld_rules``

Reviewer 2: I wonder if there's a config dump command we could provide to list the firewall rules that will be applied to each host.

Author: Added

   # Whether to install and enable firewalld.
   seed_hypervisor_firewalld_enabled: true

   # A list of zones to create. Each item is a dict containing a 'zone' item.
   seed_hypervisor_firewalld_zones: "{{ stackhpc_firewalld_zones }}"

   # A firewalld zone to set as the default. Default is unset, in which case
   # the default zone will not be changed.
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   seed_hypervisor_firewalld_default_zone: trusted

   # A list of firewall rules to apply. Each item is a dict containing
   # arguments to pass to the firewalld module. Arguments are omitted if not
   # provided, with the following exceptions:
   # - offline: true
   # - permanent: true
   # - state: enabled
   seed_hypervisor_firewalld_rules: "{{ stackhpc_firewalld_rules }}"

Kolla-Ansible configuration
---------------------------

Ensure Kolla Ansible opens up ports in firewalld for services on the public
API network:

.. code-block:: yaml
   :caption: ``etc/kayobe/kolla/globals.yml``

   enable_external_api_firewalld: true
   external_api_firewalld_zone: "{{ public_net_name | net_zone }}"

> Ensure every network in ``networks.yml`` has a zone defined. The standard

Reviewer 2: Seems like this is in the wrong section.

Author: Added a new heading above

configuration is to set the internal network zone to ``trusted`` and every
other zone to the name of the network. See
``etc/kayobe/environments/ci-multinode/networks.yml`` for a practical example.

Custom rules
------------

Custom firewalld rules can be added to ``stackhpc_firewalld_rules_extra``

The variable is a list of firewall rules to apply. Each item is a dictionary
containing arguments to pass to the firewalld module. The variable can be
defined as a group var or host var in the kayobe inventory.

The example below would enable SSH on the ``provision_oc`` network, and disable
UDP port 1000 on the ``admin_oc`` network for the Wazuh manager Infrastructure
VM:

.. code-block:: yaml
   :caption: ``etc/kayobe/inventory/group_vars/wazuh_manager/firewall``

   stackhpc_firewalld_rules_extra:
     -  service: ssh
        network: "{{ provision_oc_net_name }}"

>         zone: "{{ provision_oc_net_name | net_zone }}"

Reviewer 2: Don't need to set network, but zone is required for a custom rule.

Author: I've updated the wording here to be more specific about differences between custom and default rules

        state: enabled
     -  port: 1000/udp
        network: "{{ admin_oc_net_name }}"
        zone: "{{ admin_oc_net_name | net_zone }}"
        state: disabled

Extra rules have higher precedence than the default rules but are not
validated before being applied. Use with caution. If you need to add a custom
rule, consider adding it to the default rule list with an appropriate boolean
condition, and where possible merge your changes back into upstream SKC.

> Kolla-Ansible configuration

Reviewer 2: Duplicate of L230

Author: Durr

.. _firewall:

========
Firewall
========

StackHPC Kayobe configuration provides a standardised firewalld configuration.
The configuration uses the :kayobe-doc:`firewall
<configuration/reference/hosts.html#firewalld>` host configuration
functionality of Kayobe.

The firewall configuration is provided in
``etc/kayobe/inventory/group_vars/all/firewall``. This allows configuration
variables to be overridden on a per-group or per-host basis (which would not be
possible for an "extra variable" in ``etc/kayobe/*.yml``). This configuration
is not used by default, and must be actively opted into. This can be done as
follows:

.. code-block:: yaml
   :caption: ``etc/kayobe/controllers.yml``

   controller_firewalld_enabled: true
   controller_firewalld_rules: "{{ stackhpc_firewalld_rules }}"
   controller_firewalld_zones: "{{ stackhpc_firewalld_zones }}"
   # Predefined zones are listed here:
   # https://firewalld.org/documentation/zone/predefined-zones.html
   # Unset to leave the default zone unchanged

>    controller_firewalld_default_zone: drop

Reviewer 2: Are you planning to do enough testing to be confident in setting it to drop?

Author: I was hoping to. Whether I can get it to that point I'm not sure

.. code-block:: yaml
   :caption: ``etc/kayobe/kolla/globals.yml``

   # Open up ports in firewalld for services on the public API network.

>    enable_external_api_firewalld: true

Reviewer 2: This should move to the end to keep the flow of kayobe config.

Author: Moved
